### DISCUSSION TITLE
Using IronScheme from a User Control

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#94] [Discussions]
I recently tried to use IronScheme from a user control in my project, but attempts to drag/drop the new control onto a form raised an exception complaining about the call signature of one of the classes from the FrameworkPAL, IIRC. (It's been a couple days, and I since scrapped that approach, though I may have to revive it, so I'm posting this discussion to seek advice.)

My guess is: I probably need to install the DLLs into the GAC. The question is, assuming this is the case, as I am using DLLs I compiled from the source, which are copied into my project, and again copied into the build path I designated ... which copies of the DLLs should I register in the GAC? Does it matter?

**Comment by bcalco** [comment]
OK, I understand now why my question is stupid. LOL. Reading `package.cmd` made it clear what's happening when you (@leppie) do a full build and why it's a bad idea to rely on the various separate DLLs in the build project.

(Spoiler alert for newbies newer, or perhaps even denser, than me: ILMerge is used to roll everything needed up into the final IronScheme.dll.)

I'm still not entirely clear how compiling built-in or custom boot libraries works (the examples are always kinda vague about the presumed setup/configuration of things), but `package.cmd` did a lot to demystify most of the internals of producing the released version of the runtime and REPL EXE's. So now at least I know why using the official release, or a new release making use of the provided build scripts, is the preferred way to go, in terms of integrating with new applications.

**Comment by leppie** [comment]
The FrameworkPAL is for allowing compat with .NET Core. The library is embedded as a resource and is loaded at runtime depending on what you use.

> I'm still not entirely clear how compiling built-in or custom boot libraries works

That is part of bootstrapping. So you need IronScheme to compile IronScheme. The package part just wraps up everything into a single DLL and to remove circular dependencies. 

If you look at the tests project or the appveyor file you can see how that process works. It is best to run from the command line doing this.

You used to be able to do much easier bootstrapping while running from VS, but changes to MSBuild and the newer style project files has an issue where a newly build bootfile gets overwritten by the existing and older one.

**Comment by bcalco** [comment]
OK, so now I'm starting to think my question isn't so stupid again. I removed all the DLL references I used from the net20 release Build directory, and opted instead to nab only the IronScheme.dll from the directory into which I extracted the most recent pre-release (243), and now I'm back to this error:

![image](https://user-images.githubusercontent.com/14943929/132135719-c56536f4-dd5d-4bd1-b29e-962a896524db.png)

![image](https://user-images.githubusercontent.com/14943929/132135730-85955057-2ca1-4154-9563-0062e0ce9a10.png)

![image](https://user-images.githubusercontent.com/14943929/132135757-8f57899e-f924-4007-affc-c9c1a4012ee5.png)

So, I'm back to being somewhat confused. My confusion is somewhat augmented by the current wiki entry on embedding IronScheme, which suggests grabbing "all the DLLs" in the release directory, but there are only two, IronScheme.dll, and IronScheme.ConsoleCore.dll, and as far as I understand, they target two different .NET runtimes. The ironscheme.boot.dll is not present in the official package since it's ILMerged into IronScheme.dll.

But now that I'm referring only to that DLL in my project, I'm back to getting the errors above, which by the way are the same errors I get when I try to drag a custom user control onto the design surface at design time, if the user control in any way references an IronScheme namespace.

So it would be good to have a thorough breakdown of how to make sense of the build process, and the steps involved in embedding IronScheme with the current project structure and final build artifacts.

**Reply by bcalco** [reply]
Sorry @leppie I wrote this apparently while you were responding to my previous comment. I'll see if I can get my head around it.

**Comment by leppie** [comment]
I am not sure why that is happening. You dont have to reference the Console dll, that is just there to startup .NET Core runtime. IronScheme.dll should figure it out. 

If you grab the debug release zip, the pdb should have enough info to point it to the correct source files. 

That said, TypeInitializationExceptions are tricky to debug.

You also dont say what runtime and .NET version you are using. If you can create a small project to  repro, I can have a look.

**Reply by bcalco** [reply]
Ah yes, sorry, My project targets .NET 4.7.2. Presently I'm running strictly on Windows, cuz I want the ability to compile Scheme libraries to DLLs. I want to support ASP.NET Core as an output target (my app is going to be a kind of development environment), but need to explore the tradeoff of not being able to compile once I get the Windows only version to work, and can see if the trade off is as big in practice as it is in theory. In any case, the app itself uses WinForms controls that require the latest desktop SDK.

I will also need to create a console version of the application to run some services in separate processes, so how IronScheme.Console works is something I do need to understand fully before I embark on that.

**Reply by bcalco** [reply]
It should be reproducible in any WindowsForms app in VS 2019 that targets the latest desktop SDK. I'm not doing anything special aside from using third party components targeting the .NET desktop SDK.

**Reply by bcalco** [reply]
Also, to be clear, when I was referencing basically all the DLLs output by the IronScheme.Console project in the net20 release directory, including the FrameworkPAL and prebuilt ironscheme.boot.dll, everything "just worked." I have never yet been able to include only the IronScheme.dll, including the one from the latest pre-release (243), and NOT seen these runtime errors. I expected the fact you ILmerged the other DLLs into the IronScheme.dll that was part of the release would make the IronScheme.dll alone work without the others.

**Reply by leppie** [reply]
I will check it out.

**Reply by bcalco** [reply]
@leppie OK I think the issue was that the build directory I was outputting my application exe to was not replacing the old DLLs with the new IronScheme.dll. I just assumed "clean"-ing the solution would do that but apparently one has to manually delete third party DLLs one's self, since the IDE doesn't do this, for some reason.

Anyway I just did a fresh build having deleted all previous DLLs and now it seems my app is running as before with just the IronScheme.dll from the 243 build. I really can't believe the IDE doesn't clean third party dependency DLLs up, especially after I deleted the references to the old ones and created a new dependency on the new IronScheme.dll.

**Comment by bcalco** [comment]
Now that at least I'm using the latest greatest and properly referring to the preferred release version of IronScheme, I am going to try to use IronScheme in a user control, after having successfully used `gacutil` to install it into the GAC. I installed it from the local directory containing the IronScheme I unzipped from the prerelease binaries zip file. This directory is also in my PATH. I'll let you know how that goes. In the meantime, I'm pretty sure you should be able to reproduce that one. Just create a user control, and refer to the IronScheme namespace in it. Don't even have to use anything in it. As soon as you drop the control on a form, kaboom.

**Reply by leppie** [reply]
You dont have to install it in the GAC. I think the only reason I sign the assemblies, was to allow NGEN'ng it. A little faster start up.

> As soon as you drop the control on a form, kaboom.

Sounds like you need `IsDesignMode` checks, else it will try load everything in the control while in the designer. I havent done WinForms in like 10 years, but that used to be a sore point.  I guess extension methods are also not that friendly.

**Reply by bcalco** [reply]
I did try `IsDesignMode` and it didn't help. All that's required is including an IronScheme namespace in the `using` clause of the user control. I was quite surprised. Anyway, at some point probably today or tomorrow I'll see if putting it in the GAC helps. Based on your response, I'm tempering my expectations. :)

**Reply by leppie** [reply]
The GAC wont help. Unfortunately, IronScheme cannot have dependency on WinForms. 

What happens if you create a wrapper class/component for the `Eval` extension methods? 

Anyways, I am interested what is causing this, so I will have a peek during the week. I have a hunch, but it is too weird to admit I am even considering it :D

**Comment by leppie** [comment]
I still dont know what issue you are hitting. I just tested a simple WinForms app and cant repro. I even tried doing a custom user control. 

This is the extra code:

```
        private void button1_Click(object sender, EventArgs e)
        {
            var src = richTextBox1.Text;
            var result = src.Eval();
            richTextBox1.Text = $"{result}";
        }
```

What am I missing here?

**Reply by leppie** [reply]
I even added a call to Eval in the custom UserControl constructor. Still works fine.

**Reply by leppie** [reply]
For fun, I tried it on .NET 5, still works :D 

@bcalco the only remaining thing I can think of is that adding it to the GAC is causing a problem. 

I know there are checks to see where it is running to setup the `lib` dir. Might be that it is looking at that directory, and getting confused.

**Reply by leppie** [reply]
If you still stuck, we can setup a Skype session or something to look.

**Reply by bcalco** [reply]
Hey @leppie, sorry was out of pocket a few days. I will see if I can reproduce now that I'm using the single DLL you distribute with releases, and report back here. I did run into the issue with the lib directory being missing when the DLL is installed in the GAC and subsequently uninstalled it. So, as soon as I can focus back on this problem (which isn't stopping me, so became less of a priority to fathom out), I'll let you know what the verdict is, whether I can still reproduce the issue, or not.
